logical: add support for using new crud statements

jeffswenson opened the pull request:
This change adds a functions that call the new LDR INSERT/UPDATE/DELETE statements based on a decoded row from a replication stream.

Release note: none
Epic: CRDB-48647

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143439)
<!-- Reviewable:end -->

msbutler commented on pkg/crosscluster/logical/sql_row_writer.go:
```diff
@@ -0,0 +1,175 @@
+// sqlRowWriter is configured to write rows to a specific table and descriptor
+// version.
+type sqlRowWriter struct {
+	insert statements.Statement[tree.Statement]
```
what's the value of removing the queryBuilder abstraction, which contains a map of tableDescID to sql stmts? Perhaps this will become more apparent when i see the new row processor.

msbutler commented on pkg/crosscluster/logical/sql_row_writer.go:
```diff
@@ -0,0 +1,175 @@
+		return err
+	}
+	if err := it.Datum(func(d tree.Datum, col cdcevent.ResultColumn) error {
+		if dEnum, ok := d.(*tree.DEnum); ok {
```
why is this special enum logic only in InsertRow?

msbutler replied:
hrm, could scratch datum hydration get pushed to a helper that a bunch of these guys call? e.g. `appendToScratch(col cdceventResultColumn)`

michae2 commented on pkg/crosscluster/logical/sql_row_writer.go:
```diff
@@ -0,0 +1,175 @@
+		s.insert,
+		s.scratchDatums...,
+	)
+	return err
```
Would it be worth adding an assertion that rowsAffected is 1 after a successful insert?

michae2 reviewed (commented):
Nice job!

michae2 commented on pkg/crosscluster/logical/sql_row_writer.go:
```diff
@@ -0,0 +1,175 @@
+	// statements on initialization then reuse them, but the internal executor
+	// is scoped to a single transaction and I couldn't figure out how to
+	// maintain prepared statements across different instances of the internal
+	// executor.
```
Fixing #128911 might help.

msbutler reviewed (approved):
nice!

msbutler commented on pkg/crosscluster/logical/replication_statements.go:
```diff
@@ -185,5 +193,11 @@ func newDeleteStatement(table catalog.TableDescriptor) (tree.Statement, error) {
+func toParsedStatement(stmt tree.Statement) (statements.Statement[tree.Statement], error) {
+	// TODO(jeffswenson): do I have to round trip through the string or can I
+	// safely construct the statement directoy?
+	return parser.ParseOne(stmt.String())
```
one thing to note: we only plan to parse the statements once per distSQLFlow, not on every query execution. So the parsing overhead is probably not a big deal in the grand scheme of things.

stevendanna commented on pkg/crosscluster/logical/sql_row_writer.go:
```diff
@@ -0,0 +1,175 @@
+		return err
+	}
+
+	rowsAffected, err := txn.ExecParsed(ctx, "delete", txn.KV(),
```
Note that I think this name ends up in the application name which in turn ends up in various SQL observability surfaces, so having a common prefix really helps.

jeffswenson committed:
logical: add support for using new crud statements

This change adds a functions that call the new LDR INSERT/UPDATE/DELETE
statements based on a decoded row from a replication stream.

Release note: none
Epic: CRDB-48647

jeffswenson reviewed (commented):
I completed a virtical spike on the crud sql writer and updated this PR slightly. Instead of handling cdcevent rows, each of the functions expects tree.Datums that match the local cluster. This change is necessary because it simplifies handling read refreshes. When we read a local row, the query return datums.

jeffswenson commented:
Thanks for the reviews!

bors r+
